packages/headless/src/app/ssr-engine/ssr-engine.ts

```diff
@@ -0,0 +1,142 @@
+/**
+ * Utility functions to be used for Server Side Rendering.
+ */
+import {Middleware} from '@reduxjs/toolkit';
+import {Controller} from '../../controllers';
+import {mapObject} from '../../utils/utils';
+import {
+  SearchEngine,
+  SearchEngineOptions,
+  buildSearchEngine,
+} from '../search-engine/search-engine';
+import {EngineDefinitionBuildOptionsWithProps} from './types/build';
+import {
+  ControllerDefinitionsMap,
+  ControllersMap,
+  EngineInitialState,
+  InferControllerInitialStateMapFromDefinitions,
+  InferControllersMapFromDefinition,
+  OptionsExtender,
+} from './types/common';
+import {
+  SearchEngineDefinition,
+  SearchEngineDefinitionOptions,
+} from './types/search-engine';
+
+/**
+ * @internal
+ *
+ * Initializes a Search engine definition in SSR with given controllers definitions and search engine config.
+ * @returns Three utility functions to fetch initial state of engine in SSR, hydrate the state in CSR
+ *  and a build function that can be used for edge cases requiring more control.
+ */
+export function defineSearchEngine<
+  TControllerDefinitions extends ControllerDefinitionsMap<
+    SearchEngine,
+    Controller
+  >
+>({
+  controllers: controllerDefinitions,
+  ...engineOptions
+}: SearchEngineDefinitionOptions<TControllerDefinitions>): SearchEngineDefinition<TControllerDefinitions> {
+  const build: SearchEngineDefinition<TControllerDefinitions>['build'] = async (
+    ...[buildOptions]: Parameters<
+      SearchEngineDefinition<TControllerDefinitions>['build']
+    >
+  ) => {
+    const engine = buildSearchEngine(
+      buildOptions?.extend
+        ? await buildOptions.extend(engineOptions)
+        : engineOptions
+    );
+    const controllerOptions =
+      buildOptions && 'controllers' in buildOptions
+        ? buildOptions.controllers
+        : null;
+    const controllers = controllerDefinitions
+      ? mapObject(controllerDefinitions, (definition, key) =>
+          'build' in definition
+            ? definition.build(engine)
+            : definition.buildWithProps(
+                engine,
+                controllerOptions?.[key as keyof typeof controllerOptions]
+              )
+        )
+      : {};
+    return {
+      engine,
+      controllers:
+        controllers as InferControllersMapFromDefinition<TControllerDefinitions>,
+    };
+  };
+
+  const fetchInitialState: SearchEngineDefinition<TControllerDefinitions>['fetchInitialState'] =
+    (
+      ...[executeOptions]: Parameters<
+        SearchEngineDefinition<TControllerDefinitions>['fetchInitialState']
+      >
+    ) =>
+      new Promise<
+        EngineInitialState<
+          {type: string},
+          InferControllerInitialStateMapFromDefinitions<TControllerDefinitions>
+        >
+      >((resolve, reject) => {
+        let initialControllers: ControllersMap;
```

Comment:
Managed to fix the eslint no-async-promise-executor warn by refactoring to extract the controller out. Don't love this but not sure there is a better solution 🤷‍♂️.